cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java

```diff
@@ -1516,25 +1529,25 @@ private static void templates(LexerfulGrammarBuilder b) {
 
     b.rule(templateArgument).is(
       b.firstOf(
-        typeId,
-        
+        b.sequence(typeId, b.next(b.firstOf(">", ",", "..."))),
+        b.sequence(typenameSpecifier, b.next(b.firstOf(">", ",", "..."))), // seen in gnu system headers
         // FIXME: workaround to parse stuff like "carray<int, 10>"
         // actually, it should be covered by the next rule (constantExpression)
         // but it doesnt work because of ambiguity template syntax <--> relationalExpression
-        shiftExpression,
-        constantExpression,
-        idExpression
+        b.sequence(shiftExpression, b.next(b.firstOf(">", ","))),
+        b.sequence(constantExpression, b.next(b.firstOf(">", ",")))
+//        idExpression
       )
```

Comment:
@ametselaar Thanks a lot for providing the solution. 

Grammar and order is slightly different compared to standard. Especially id-expressions seems to be not there?
Could you give a short hint why please.
```
template-argument:
   constant-expression
   type-id
   id-expression
```

```
id-expression:
   unqualified-id
   qualified-id
unqualified-id:
   identifier
   operator-function-id
   conversion-function-id
   literal-operator-id
   ~ class-name
   ~ decltype-specifier
   template-id
qualified-id:
   nested-name-specifier templateopt unqualified-id
````